File: AssistantPlus/Application/Utility/ApplicationExceptionManager.cs
#pragma warning disable 0,

using Twinning.AssistantPlus;

namespace Twinning.AssistantPlus.Utility {

	public class ApplicationExceptionManager {

		#region singleton

		public static readonly ApplicationExceptionManager Instance = new ();

		#endregion

		#region constructor

		private Boolean mInitialized;

		private Func<Exception, Task>? mHandler;

		// ----------------

		private ApplicationExceptionManager(
		) {
			this.mInitialized = false;
			this.mHandler = null;
			return;
		}

		#endregion

		#region utility

		public async Task Initialize(
			App application
		) {
			AssertTest(!this.mInitialized);
			application.UnhandledException += async (_, args) => {
				args.Handled = true;
				this.mHandler?.Invoke(args.Exception);
				return;
			};
			TaskScheduler.UnobservedTaskException += async (_, args) => {
				args.SetObserved();
				this.mHandler?.Invoke(args.Exception);
				return;
			};
			this.mInitialized = true;
			return;
		}

		public async Task Listen(
			Func<Exception, Task>? handler
		) {
			AssertTest(this.mInitialized);
			this.mHandler = handler;
			return;
		}

		// ----------------

		public Task WrapTask(
			Task task
		) {
			return task.ContinueWith((it) => {
				if (it.Exception?.InnerException != null) {
					this.mHandler?.Invoke(it.Exception.InnerException);
				}
				return;
			});
		}

		#endregion

	}

}

File: Assistant/lib/utility/application_exception_manager.dart
import '/common.dart';
import 'package:flutter/widgets.dart';

// ----------------

class ApplicationExceptionManager {

  // #region singleton

  static final ApplicationExceptionManager instance = ._();

  // #endregion

  // #region constructor

  Boolean _initialized;

  Future<Void> Function(Object exception, StackTrace? stack)? _handler;

  // ----------------

  ApplicationExceptionManager._(
  ) :
    _initialized = false,
    _handler = null;

  // #endregion

  // #region utility

  Future<Void> initialize(
  ) async {
    assertTest(!this._initialized);
    WidgetsBinding.instance.platformDispatcher.onError = (exception, stack) {
      this._handler?.call(exception, stack);
      return true;
    };
    FlutterError.onError = (details) {
      FlutterError.presentError(details);
      this._handler?.call(details.exception, details.stack);
      return;
    };
    this._initialized = true;
    return;
  }

  Future<Void> listen(
    Future<Void> Function(Object exception, StackTrace? stack)? handler,
  ) async {
    assertTest(this._initialized);
    this._handler = handler;
    return;
  }

  // #endregion

}
